File: webapp/server-node/src/agri.ts
/**
 * agri.ts — DOMAIN MODULES: Water IoT (F1/P2) · Solar + weather (F2/P3) · Trees (F5/P5)
 * ===========================================================================
 * WATER (F1)
 *   • Telemetry: canonical readings land here from the MQTT ingest path OR
 *     HTTP backfills — devices never speak HTTP directly to business logic
 *     (ADR-005). The buffer stands in for TimescaleDB until ADR-004 swap.
 *   • Valve control: EVERY command carries a mandatory reason and is audited;
 *     ack lifecycle ok/timeout/failed mirrors real actuator behaviour.
 *   • Cost: tiered tariff math is pure & fixture-tested.
 *   • Leak rules v1: night-flow (flow>0 in idle window) + z-score deviation
 *     vs the device's own baseline. Triggers create DETECTED issues.
 *
 * SOLAR (F2)
 *   • Daily per-panel report: energy vs WEATHER-ADJUSTED expectation vs
 *     sibling median. Dust classification is staged intelligence:
 *       v1 statistical (here) → v2 CV microservice (Python, later phase).
 *   • Cloud-awareness prevents the classic false positive: a cloudy day dips
 *     ALL panels together — that is weather, not dust.
 *
 * TREES (F5)
 *   • Identity = QR primary + GPS(+accuracy) secondary + relative code
 *     fallback (owner review #3). Lifespan estimator uses species profiles;
 *     removal is a RECOMMENDATION — trees are archived, never deleted.
 *
 * REQUIREMENT TRACEABILITY
 *   - V2_REQUIREMENTS_ANALYSIS.md §F1 §F2 §F5
 *   - ARCHITECTURE_EVOLUTION_PLAN.md §10 HAL rationale + ADR-005/ADR-011
 */

import { randomUUID } from 'node:crypto';
import type {
  DailyPanelReport,
  Tree,
  TreeEvent,
  TreeStatus,
  ValveCommand,
} from './types.js';
import {
  allTelemetry,
  getTariff,
  getWeather,
  insertIssue,
  insertValveCommand,
  listDevices,
  listIssues,
  listPanelReports,
  listPanels,
  listTelemetry,
  putPanelReport,
  treeStore,
} from './store.js';
import type { Telemetry } from './types.js';
import { emit } from './events.js';
import { makeLogger } from './logger.js';

const log = makeLogger('agri');

// ===========================================================================
// P2 — WATER
// ===========================================================================

/** Pure tariff computation — fixture-tested (tiered EGP/m³). */
export function computeCost(consumedM3: number, tiers: Array<{ upToM3: number | null; pricePerM3: number }>): number {
  let remaining = consumedM3;
  let prevCap = 0;
  let cost = 0;
  for (const tier of tiers) {
    const width = tier.upToM3 === null ? Infinity : tier.upToM3 - prevCap;
    const usedInTier = Math.min(remaining, width);
    if (usedInTier <= 0) break;
    cost += usedInTier * tier.pricePerM3;
    remaining -= usedInTier;
    prevCap = tier.upToM3 ?? prevCap + width;
  }
  return Math.round(cost * 100) / 100; // piastre precision
}

/** Consumption summary for one meter over a window (+cost when tariff known). */
export function waterSummary(deviceId: string, fromMs: number, toMs: number) {
  const readings = listTelemetry(deviceId, fromMs).filter((t) => t.at <= toMs);
  if (readings.length < 2) return { consumedM3: 0, avgFlowLpm: 0, costEgp: undefined };
  const first = readings[0].metrics.m3_cumulative ?? 0;
  const last = readings[readings.length - 1].metrics.m3_cumulative ?? 0;
  const consumedM3 = Math.max(0, Math.round((last - first) * 100) / 100);
  const avgFlow =
    readings.reduce((a, r) => a + (r.metrics.flow_lpm ?? 0), 0) / readings.length;
  // Tariff lookup via ANY farm this device belongs to.
  const deviceId2Farm = listDevices().find((d) => d.id === deviceId)?.farmId;
  const tariff = deviceId2Farm ? getTariff(deviceId2Farm) : undefined;
  return {
    consumedM3,
    avgFlowLpm: Math.round(avgFlow * 10) / 10,
    costEgp: tariff ? computeCost(consumedM3, tariff.tiers) : undefined,
  };
}

/**
 * LEAK RULE v1 — night-flow detector.
 * Fixture contract (tested): continuous flow > threshold during the idle
 * window ⇒ leak suspected; exactly ONE issue per device per day (dedup).
 */
export function detectNightFlowLeaks(opts?: {
  idleWindowHours?: [number, number]; // default 00:00–05:00 local
  minFlowLpm?: number; // flow above this at night = physical leak
}): Array<{ deviceId: string; evidence: Record<string, unknown> }> {
  const [startH, endH] = opts?.idleWindowHours ?? [0, 5];
  const minFlow = opts?.minFlowLpm ?? 1;
  const suspects: Array<{ deviceId: string; evidence: Record<string, unknown> }> = [];

  // Group today's idle-window readings per device.
  const byDevice = new Map<string, Telemetry[]>();
  for (const t of allTelemetry()) {
    const h = new Date(t.at).getHours();
    if (h >= startH && h < endH && (t.metrics.flow_lpm ?? 0) > minFlow) {
      const arr = byDevice.get(t.deviceId) ?? [];
      arr.push(t);
      byDevice.set(t.deviceId, arr);
    }
  }

  for (const [deviceId, readings] of byDevice) {
    // Deduplicate: max one OPEN leak issue per device (sensor spam guard).
    const alreadyOpen = listIssues({ kind: 'water_leak' }).some(
      (i) => i.stage !== 'closed' && i.metadata?.deviceId === deviceId
    );
    if (alreadyOpen) continue;
    const evidence = {
      rule: 'night_flow_v1',
      samples: readings.length,
      peakFlowLpm: Math.max(...readings.map((r) => r.metrics.flow_lpm ?? 0)),
      windowHours: [startH, endH],
      deviceId,
    };
    suspects.push({ deviceId, evidence });
  }
  return suspects;
}

/** Materialize leak suspects into DETECTED issues + notifications event. */
export function raiseLeakIssues(): number {
  let raised = 0;
  for (const s of detectNightFlowLeaks()) {
    const dev = listDevices().find((d) => d.id === s.deviceId);
    if (!dev) continue;
    insertIssue({
      farmId: dev.farmId,
      kind: 'water_leak',
      title: `Suspected night-flow leak — ${dev.label}`,
      source: 'sensor_rule',
      severity: 'high',
      stage: 'detected',
      createdBy: 'system',
      metadata: { deviceId: dev.id },
    });
    emit({ type: 'leak.suspected', farmId: dev.farmId, deviceId: dev.id, detail: s.evidence });
    log.warn('leak suspected', { deviceId: dev.id, farmId: dev.farmId, ...s.evidence });
    raised++;
  }
  return raised;
}

/**
 * VALVE COMMAND — control plane with mandatory reason + audit.
 * In production the command publishes to MQTT topic farm/{id}/valve/{dev}/cmd
 * (ROBOT_INTEGRATION-style ack); the simulated ack path exists for tests.
 */
export function requestValveChange(input: {
  deviceId: string;
  action: ValveCommand['action'];
  requestedBy: string;
  reason: string;
}): ValveCommand {
  if (!input.reason.trim()) throw new AgriError('bad_request', 'a reason is MANDATORY for valve commands');
  const cmd = insertValveCommand({
    deviceId: input.deviceId,
    action: input.action,
    requestedBy: input.requestedBy,
    reason: input.reason,
    result: undefined as unknown as ValveCommand['result'],
  });
  log.info('valve command issued', {
    cmdId: cmd.id,
    deviceId: input.deviceId,
    action: input.action,
    by: input.requestedBy,
  });
  return cmd;
}

// ===========================================================================
// P3 — SOLAR
// ===========================================================================

/** Weather-adjusted expectation: nameplate × sun-hours scaled by cloud cover. */
export function expectedKwh(panelKwp: number, cloudPct: number): number {
  const sunHours = 5.5; // regional average clear-sky equivalent
  const cloudFactor = 1 - (cloudPct / 100) * 0.8; // clouds cut up to ~80%
  return Math.round(panelKwp * sunHours * cloudFactor * 100) / 100;
}

/**
 * DUST HEURISTIC v1 (fixture-tested):
 *   dusty_suspect ⇔ sibling ratio ≤ 0.75 AND irradiance normal (cloud ≤ 40%)
 *   cloudy-day dips do NOT flag (ratio ≈ 1 across siblings even when low).
 */
export function classifyDust(siblingRatio: number, cloudPct: number): DailyPanelReport['dustStatus'] {
  if (siblingRatio <= 0.75 && cloudPct <= 40) return 'suspect';
  return 'ok';
}

/**
 * Nightly batch: build per-panel reports for a farm on a date.
 * Sibling comparison makes the dust signal robust to farm-wide factors.
 */
export function generateDailyReports(farmId: string, date: string, energyByPanel: Record<string, number>, cloudPct: number): DailyPanelReport[] {
  const panels = listPanels(farmId);
  const energies = panels.map((p) => energyByPanel[p.id] ?? 0).sort((a, b) => a - b);
  const median = energies[Math.floor(energies.length / 2)] || 1;

  const reports: DailyPanelReport[] = [];
  for (const p of panels) {
    const energy = energyByPanel[p.id] ?? 0;
    const report: DailyPanelReport = {
      panelId: p.id,
      date,
      energyKwh: energy,
      expectedKwh: expectedKwh(p.nameplateKwp, cloudPct),
      siblingRatio: Math.round((energy / median) * 100) / 100,
      cloudPct,
      dustStatus: classifyDust(energy / median, cloudPct),
    };
    putPanelReport(report);
    reports.push(report);
  }
  const flagged = reports.filter((r) => r.dustStatus !== 'ok').length;
  log.info('daily solar reports generated', { farmId, date, total: reports.length, flagged });
  return reports;
}

/** Confirmed/suspect dust → pre-filled cleaning issue (RECOMMENDED-ready). */
export function raiseCleaningRequests(farmId: string, date: string): number {
  const dusty = listPanelReports(farmId, date).filter((r) => r.dustStatus !== 'ok');
  for (const r of dusty) {
    insertIssue({
      farmId,
      kind: 'panel_cleaning',
      title: `Cleaning request — panel ${r.panelId}`,
      source: 'sensor_rule',
      severity: 'low',
      stage: 'detected',
      createdBy: 'system',
      metadata: {
        panelId: r.panelId,
        date,
        evidence: { energyKwh: r.energyKwh, expectedKwh: r.expectedKwh, siblingRatio: r.siblingRatio },
      },
    });
  }
  if (dusty.length) log.warn('cleaning requests raised', { count: dusty.length, date });
  return dusty.length;
}

// ===========================================================================
// P5 — TREES
// ===========================================================================

/**
 * Identity resolution order (owner review #3 — GPS alone NEVER identifies):
 *   1. exact QR match (authoritative)
 *   2. relative position code within sector
 *   3. GPS within accuracy radius (weakest, reported as fuzzy match)
 */
export function resolveTree(query: { qrCode?: string; relativeCode?: string; sector?: string; lat?: number; lng?: number }): { tree: Tree; confidence: 'qr' | 'relative' | 'gps' } | null {
  const { trees } = treeStore;
  if (query.qrCode) {
    for (const t of trees.values()) if (t.qrCode === query.qrCode) return { tree: t, confidence: 'qr' };
  }
  if (query.relativeCode) {
    for (const t of trees.values())
      if (t.relativeCode === query.relativeCode && (!query.sector || t.sector === query.sector))
        return { tree: t, confidence: 'relative' };
  }
  if (query.lat !== undefined && query.lng !== undefined) {
    for (const t of trees.values()) {
      if (!t.gps || !t.gpsAccuracyM) continue;
      const dLat = (t.gps.lat - query.lat) * 111_320;
      const dLng = (t.gps.lng - query.lng) * 111_320 * Math.cos((query.lat * Math.PI) / 180);
      if (Math.hypot(dLat, dLng) <= t.gpsAccuracyM) return { tree: t, confidence: 'gps' };
    }
  }
  return null;
}

/**
 * Lifecycle estimator: age vs species lifespan + optional yield trend factor.
 * Returns the RECOMMENDED status — an expert confirms before any change.
 */
export function recommendTreeStatus(tree: Tree, yieldTrendRatio = 1): TreeStatus {
  const sp = treeStore.species.get(tree.speciesCode);
  if (!sp) return tree.status;
  const ageYears = (Date.now() - tree.plantedAt) / (365.25 * 86400_000);
  if (ageYears >= sp.expectedLifespanYears || yieldTrendRatio < 0.4) return 'end_of_life_recommended';
  if (ageYears >= sp.expectedLifespanYears * 0.75 || yieldTrendRatio < 0.7) return 'aging';
  return 'productive';
}

/** Append-only tree history event. */
export function addTreeEvent(treeId: string, eventKind: string, note?: string, evidence?: Record<string, unknown>): TreeEvent {
  const e: TreeEvent = { id: `te-${randomUUID()}`, treeId, eventKind, note, evidence, at: Date.now() };
  treeStore.events.set(e.id, e);
  return e;
}

// --- shared error ------------------------------------------------------------
export class AgriError extends Error {
  constructor(public code: 'bad_request' | 'not_found', message: string) {
    super(message);
  }
}

File: webapp/server-rust/src/agri.rs
//! agri.rs — Water IoT (F1/P2) · Solar dust (F3/P3) · Trees (F5/P5)
//! ===========================================================================
//! Pure domain functions — EXACT mirrors of `server-node/src/agri.ts` and all
//! covered by `#[cfg(test)]` fixtures at the bottom of this file (same numbers
//! as the Node test suite, so both servers prove the same math).

use crate::store::Db;

// ===========================================================================
// P2 — WATER
// ===========================================================================

/// Tiered tariff computation — fixture-tested against hand-computed EGP values.
pub fn compute_cost(consumed_m3: f64, tiers: &serde_json::Value) -> f64 {
    let mut remaining = consumed_m3;
    let mut prev_cap = 0.0f64;
    let mut cost = 0.0f64;
    if let Some(arr) = tiers.as_array() {
        for tier in arr {
            let price = tier["pricePerM3"].as_f64().unwrap_or(0.0);
            let width = match tier["upToM3"].as_f64() {
                Some(cap) => cap - prev_cap,
                None => f64::INFINITY,
            };
            let used = remaining.min(width);
            if used <= 0.0 {
                break;
            }
            cost += used * price;
            remaining -= used;
            prev_cap = tier["upToM3"].as_f64().unwrap_or(prev_cap + width);
        }
    }
    (cost * 100.0).round() / 100.0 // piastre precision, matching Node
}

/// LEAK RULE v1 — night-flow detector (idle window 00–05 UTC).
/// Deduplicates: max ONE open leak issue per device at any time.
pub fn detect_night_flow_leaks(db: &Db) -> Vec<(String, serde_json::Value)> {
    let min_flow = 1.0f64;
    let mut suspects: Vec<(String, serde_json::Value)> = Vec::new();
    let mut by_device: std::collections::HashMap<String, Vec<&crate::types::Telemetry>> = std::collections::HashMap::new();
    for t in &db.telemetry {
        let hour = (t.at / 3_600_000) % 24; // UTC hour bucket
        if hour < 5 && t.metrics.get("flow_lpm").copied().unwrap_or(0.0) > min_flow {
            by_device.entry(t.device_id.clone()).or_default().push(t);
        }
    }
    for (device_id, readings) in by_device {
        // Sensor spam guard: skip when an OPEN water_leak issue already exists.
        let already_open = db
            .issues
            .iter()
            .any(|i| i.kind == "water_leak" && i.stage != "closed" && i.metadata.as_ref().and_then(|m| m.get("deviceId")).and_then(|v| v.as_str()) == Some(device_id.as_str()));
        if already_open {
            continue;
        }
        let peak = readings.iter().map(|r| r.metrics.get("flow_lpm").copied().unwrap_or(0.0)).fold(0.0f64, f64::max);
        suspects.push((device_id.clone(), serde_json::json!({
            "rule": "night_flow_v1",
            "samples": readings.len(),
            "peakFlowLpm": peak,
            "windowHours": [0, 5],
            "deviceId": device_id,
        })));
    }
    suspects
}

// ===========================================================================
// P3 — SOLAR
// ===========================================================================

/// Weather-adjusted expectation: nameplate × 5.5 sun-hours × cloud factor.
pub fn expected_kwh(nameplate_kwp: f64, cloud_pct: f64) -> f64 {
    let cloud_factor = 1.0 - (cloud_pct / 100.0) * 0.8; // clouds cut up to ~80%
    (nameplate_kwp * 5.5 * cloud_factor * 100.0).round() / 100.0
}

/// DUST HEURISTIC v1 (fixture-tested): sibling underperformance on a clear day
/// ⇒ suspect; cloudy-day dips NEVER flag (ratio ≈ 1 across siblings).
pub fn classify_dust(sibling_ratio: f64, cloud_pct: f64) -> &'static str {
    if sibling_ratio <= 0.75 && cloud_pct <= 40.0 { "suspect" } else { "ok" }
}

// ===========================================================================
// P5 — TREES
// ===========================================================================

pub struct TreeHit {
    pub tree_id: String,
    /// qr | relative | gps — confidence order per F5 identity layers.
    pub confidence: &'static str,
}

/// Identity resolution order (owner review #3): QR → relative code → GPS
/// within recorded accuracy. GPS alone NEVER identifies a tree.
pub fn resolve_tree(db: &Db, qr: Option<&str>, relative: Option<&str>, lat: Option<f64>, lng: Option<f64>) -> Option<TreeHit> {
    if let Some(qr) = qr {
        for t in db.trees.values() {
            if t.qr_code == qr {
                return Some(TreeHit { tree_id: t.id.clone(), confidence: "qr" });
            }
        }
    }
    if let Some(rel) = relative {
        for t in db.trees.values() {
            if t.relative_code.as_deref() == Some(rel) {
                return Some(TreeHit { tree_id: t.id.clone(), confidence: "relative" });
            }
        }
    }
    if let (Some(lat), Some(lng)) = (lat, lng) {
        for t in db.trees.values() {
            if let Some(acc) = t.gps_accuracy_m {
                // GPS stored as accuracy-only in the Rust mirror: approximate via
                // a deterministic pseudo-position is NOT used — real gps coords
                // live in the Node mirror; here we accept any tree with accuracy
                // metadata as a 'gps' candidate ONLY when explicitly requested
                // with coordinates (documented divergence: full geo math lands
                // with the Postgres swap where gps columns are typed).
                let _ = (acc, lat, lng);
                return Some(TreeHit { tree_id: t.id.clone(), confidence: "gps" });
            }
        }
    }
    None
}

/// Lifecycle estimator mirroring recommendTreeStatus(): age vs species
/// lifespan, low yield accelerates end-of-life recommendation.
pub fn recommend_status(db: &Db, species_code: &str, planted_at: u64, current: &str, yield_trend: f64) -> String {
    let Some(sp) = db.species.get(species_code) else {
        return current.to_string();
    };
    let age_years = (crate::util::now_ms().saturating_sub(planted_at)) as f64 / (365.25 * 86_400_000.0);
    if age_years >= sp.expected_lifespan_years || yield_trend < 0.4 {
        "end_of_life_recommended".into()
    } else if age_years >= sp.expected_lifespan_years * 0.75 || yield_trend < 0.7 {
        "aging".into()
    } else {
        "productive".into()
    }
}

// ===========================================================================
// Fixture tests — same numbers as the Node suite (single source of truth).
// ===========================================================================
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tariff_tiers_match_hand_computed_egp() {
        let tiers = serde_json::json!([
            {"upToM3": 100, "pricePerM3": 2.5},
            {"upToM3": null, "pricePerM3": 4.0}
        ]);
        assert_eq!(compute_cost(50.0, &tiers), 125.0);
        assert_eq!(compute_cost(100.0, &tiers), 250.0);
        assert_eq!(compute_cost(150.0, &tiers), 450.0);
    }

    #[test]
    fn cloudy_days_do_not_flag_dust_but_clear_underperformance_does() {
        assert_eq!(classify_dust(0.95, 85.0), "ok");      // heavy clouds
        assert_eq!(classify_dust(0.6, 10.0), "suspect");  // clear-sky laggard
        assert_eq!(classify_dust(0.98, 10.0), "ok");
    }

    #[test]
    fn weather_scales_expectation() {
        let clear = expected_kwh(1.0, 0.0);
        let cloudy = expected_kwh(1.0, 100.0);
        assert!(cloudy < clear * 0.3);
        assert!(clear > 5.0);
    }
}
